src/main/java/com/faforever/api/map/MapService.java

```diff
@@ -0,0 +1,348 @@
+package com.faforever.api.map;
+
+import com.faforever.api.config.FafApiProperties;
+import com.faforever.api.content.ContentService;
+import com.faforever.api.data.domain.Map;
+import com.faforever.api.data.domain.MapVersion;
+import com.faforever.api.data.domain.Player;
+import com.faforever.api.error.ApiException;
+import com.faforever.api.error.Error;
+import com.faforever.api.error.ErrorCode;
+import com.faforever.api.utils.JavaFxUtil;
+import com.faforever.api.utils.PreviewGenerator;
+import com.faforever.api.utils.Unzipper;
+import com.faforever.api.utils.Zipper;
+import javafx.scene.image.Image;
+import lombok.Data;
+import lombok.SneakyThrows;
+import org.luaj.vm2.LuaValue;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.FileSystemUtils;
+
+import javax.inject.Inject;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.zip.ZipInputStream;
+import java.util.zip.ZipOutputStream;
+
+import static com.faforever.api.utils.LuaUtil.loadFile;
+import static com.github.nocatch.NoCatch.noCatch;
+
+@Service
+public class MapService {
+  private static final String[] REQUIRED_FILES = new String[]{
+      ".scmap",
+      "_save.lua",
+      "_scenario.lua",
+      "_script.lua"};
+  public static final Charset MAP_CHARSET = StandardCharsets.ISO_8859_1;
+  private final FafApiProperties fafApiProperties;
+  private final MapRepository mapRepository;
+  private final ContentService contentService;
+
+  @Inject
+  public MapService(FafApiProperties fafApiProperties, MapRepository mapRepository, ContentService contentService) {
+    this.fafApiProperties = fafApiProperties;
+    this.mapRepository = mapRepository;
+    this.contentService = contentService;
+  }
+
+  @Transactional
+  public void uploadMap(byte[] mapData, String mapFilename, Player author, boolean isRanked) throws IOException {
+    if (author == null) {
+      throw new IllegalStateException("author cannot be null");
+    }
+    if (mapData.length <= 0) {
+      throw new IllegalStateException("mapData is empty");
+    }
```

Comment:
This is pure Application error, file data is required in the controller and spring will internal handle it: 

![image](https://cloud.githubusercontent.com/assets/1217999/22431582/55eeb70e-e712-11e6-8387-45329425e7da.png)